File: v1/src/main/java/com/api/v1/domain/entities/Purchase.java
package com.api.v1.domain.entities;

import com.api.v1.utils.book.SaleTaxRatesUtil;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.ZonedDateTime;

@Document(collection = "v1_purchases")
public record Purchase(
        @NotNull Book book,
        @NotNull Customer customer,
        double bookPrice,
        @NotBlank String state,
        Double saleTax,
        Double finalPrice,
        String createdAt
) {

    public Purchase(Book book, Customer customer, String state) {
        this(
                book,
                customer,
                book.getPrice(),
                state,
                SaleTaxRatesUtil.map(state),
                (1 + SaleTaxRatesUtil.map(state)) * book.getPrice(),
                ZonedDateTime.now().toString()
        );
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/purchases/domain/Purchase.kt
package com.api.v2.purchases.domain

import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.math.BigInteger
import java.time.Instant
import java.time.ZoneId
import java.util.*

@Table("purchases")
data class Purchase(
    @Id
    var id: UUID?,
    var orderNumber: BigInteger,
    var customerId: UUID,
    var bookId: UUID,
    var price: Double,
    var finalPrice: Double,
    var createdAt: Instant,
    var creationZoneId: ZoneId
)
